[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112861
* #114687
* #115270
* #115174
* #115072
* #114606
* #114319
* #114075
* #113932
* #113891
* #113959
* #113890
* #113570
* #113569
* #113558
* #113399
* __->__ #113367
* #113366
* #113365

Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

kadeng committed:
[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend

Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_epilogue_gen.py:
```diff
@@ -107,6 +122,23 @@ def ir_to_evt_string(
 
+    @staticmethod
+    def create_pre_fused_addmm_evt_type() -> str:
+        """returns the name of the ADDMM EVT type which has been declared like this:
```
Super-nit: let's start from upper-case R.

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_epilogue_gen.py:
```diff
@@ -208,6 +243,9 @@ def getvalue(self, result) -> str:
         self.output.writeline(f"using {self.evt_type_name} = {dtype_converted_expr};")
         return self.output.getvalue()
 
+    def _op_pre_fused_expr(self, expr):
+        return expr
+
```
Nit: move this up to other `_op_*`s?

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_epilogue_gen.py:
```diff
@@ -236,24 +278,31 @@ def __init__(self, accumulator_node_name: str):
-        with virtualized.V.set_ops_handler(formatter), patch.object(
+        result = pre_fused_evt_args
+        if (pre_fused_evt_args is None) and (
+            (epilogue_nodes is None) or len(epilogue_nodes) == 0
```
Can `epilogue_nodes` be `None`? Need to update the type hint above if so.

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_epilogue_gen.py:
```diff
@@ -236,24 +278,31 @@ def __init__(self, accumulator_node_name: str):
         )
-
-        with virtualized.V.set_ops_handler(formatter), patch.object(
+        result = pre_fused_evt_args
```
I believe, it would be cleaner to move this line (using `pre_fused_evt_args`) after the following `if` statement. A bit confusing otherwise.

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_epilogue_gen.py:
```diff
@@ -274,6 +323,24 @@ def ir_to_evt_argument_string(
+        return """
+        {  // ADDMM Arguments: ternary op : beta * C + (alpha * acc)
+          {{static_cast<ElementAcc>(%f)}}, // leaf op+args : beta
+          {},               // leaf op+args : C
```
Are `C` and `acc` not specified in the argument structure?

aakhundov commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -337,17 +337,28 @@ def render_evt_epilogue_declaration(
         template_output_node_name: str,
         evt_type_name: str,
         epilogue_nodes: List[IRNode],
+        Bias: Optional[Buffer] = None,
```
Nit: does this need to be upper-case?

aakhundov commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -337,17 +337,28 @@ def render_evt_epilogue_declaration(
         op: "cutlass_library.gemm_op.GemmOperation",  # type: ignore[name-defined]
         output_buffer_name: str,
         epilogue_nodes: Optional[List[IRNode]] = None,
+        Bias: Optional[Buffer] = None,
```
ditto

aakhundov commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -452,7 +470,7 @@ def fix_op_layout(
             return op
         log.warning(
-            f"Cutlass GEMM Layout change: Input and/or output layouts have changed between autotuning and call to render on {self}. Applying workaround. This can lead to suboptimal performance."
+            f"Cutlass GEMM Layout change: Input and/or output layouts have changed between autotuning and call to render on {self}. Applying workaround. This can lead to suboptimal performance." # noqa: G004, B950
```
Nit: could still break this into multiple strings and avoid `# noqa`.

aakhundov reviewed (changes requested):
Thanks, @kadeng, LGTM overall! Left a few minor comments. Some CI errors look legit. Also there are linter errors.

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor CUTLASS backend] Cutlass EVT Addmm support for Inductor Cutlass backend"


Adding support for pre-fused ADDMM using Cutlass EVT functors, which
would allow to fuse EVT epilogues with Cutlass GEMM ops, which
implement the aten.addmm operation.

Note:

Currently this is not of much practical use, since at some
point during torch.compile, most non-standalone aten.addmm ops
seem to be decomposed into aten.mm + further ops before
lowering.

Test Plan:
 - CI
 - Added unit test within test/inductor/test_max_autotune.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng commented:
Moved to a (draft) feature branch, see https://github.com/pytorch/pytorch/pull/115919
